src/lib.rs

```diff
@@ -1237,6 +1392,53 @@ pub async fn build(docker: &Docker, config: &Config) -> Result<(), Error> {
         }
     }
 
+    // Generate {appservice}.yml for appservices running on the host.
+    let generate_appservice_yml =
+        |appservice: &AppServiceConfig, host: &str, path: &Path| -> Result<(), Error> {
+            let mut config = appservice.clone();
+            config.url.set_host(Some(host))?;
+            let writer = std::fs::File::create(&path)
+                .with_context(|| format!("Failed to create file {:?}", path))?;
+            serde_yaml::to_writer(writer, &config)
+                .with_context(|| format!("Failed to write file {:?}", path))?;
+            Ok(())
+        };
+
+    let guest_appservices_root = synapse_root.join("appservices");
+    let host_appservices_root = config.generated_user_files_dir().join("appservices");
+
+    // The (guest) Synapse sees them as running on `host.docker.internal`,
+    // while the (host) AppService see themselves as running on `localhost`.
+    for appservice in &config.appservices.host {
+        // Guest file
+        let guest_path = guest_appservices_root
+            .join(appservice.name.as_ref())
+            .with_extension("yml");
+        generate_appservice_yml(appservice, "host.docker.internal", &guest_path)?;
+        // Host file
+        let host_path = host_appservices_root
+            .join(appservice.name.as_ref())
+            .with_extension("yml");
+        generate_appservice_yml(appservice, "localhost", &host_path)?;
+        eprintln!(
+            "Configuration file for appservice {} has been generated at {:?}",
+            appservice.name, host_path
+        );
```

Comment:
```
thread 'main' panicked at 'Error in `build`: Failed to create file "/tmp/mx-tester/mjolnir/synapse/appservices/mjolnir.yml"

Caused by:
    No such file or directory (os error 2)', src/main.rs:250:47
```

I'm not sure why this didn't work

```
appservices:
  # Optional. Define Matrix Application Services to be connected
  # with your homeserver. For more details on AppServices, see
  # https://spec.matrix.org/v1.2/application-service-api .
  host:
    # Optional. Application Services whose process is running on
    # the host.
    - name: mjolnir
      # String. The (local) URL at which to communicate with the AppService,
      # including a port number, e.g. "http://localhost:8765".
      # Note that the host is actually IGNORED.
      url: "http://host.docker.internal:9000"
      as_token:
        # Optional. String. If specified, a token used by the homeserver
        # when sending messages to the AppService.
        # By default, generate a random token automatically during `build`.
        # See the generated appservice file to determine its value.
        whatever
      hs_token:
        # Optional. String. If specified, a token used by the AppService
        # when sending messages to the homeserver.
        # By default, generate a random token automatically during `build`.
        # See the generated appservice file to determine its value.
        whatever
      sender_localpart:
        # String. The localpart of the userid of the user that will be
        # used when this AppService sends messages.
        mjolnir
      namespaces:
        # The namespaces controlled or monitored by this AppService.
        users:
          # Optional. Empty by default.
          - regex:
              # A set of users to control or monitor.
              # Regex are matched against the full user id, including sigil
              # and homeserver name.
              '@mjolnir_.*'
            exclusive:
              # Optional. Boolean. False by default.
              # If `true`, the AppService is in charge of all users whose
              # id matches the regex. Otherwise, the AppService will only
              # be informed of the actions of these users.
              true
```

Reply:
For some reason, it works better if I actually create the directory.
sigh
Added a minimal test.